Add citool command for generating a test dashboard

Kobzol opened the pull request:
This PR implements an initial version of a test suite dashboard, which shows which tests were executed on CI and which tests were ignored. This was discussed [here](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Bsteering.5D.202025-04-11.20Dealing.20with.20disabled.20tests/with/512799036). The dashboard is still quite bare-bones, but I think that it could already be useful.

The next step is to create a job index, similarly to the post-merge report, and link from the individual tests to the job that executed them.

You can try it locally like this:
```bash
$ cargo run --manifest-path src/ci/citool/Cargo.toml --release \
    -- test-dashboard 38c560ae681d5c0d3fd615eaedc537a282fb1086 --output-dir dashboard
```
and then open `dashboard/index.html` in a web browser.

CC @wesleywiser

r? @jieyouxu

Kobzol committed:
Make `parent` in `download_auto_job_metrics` optional

Kobzol committed:
Extract function for normalizing path delimiters to `utils`

Kobzol committed:
Add command to `citool` for generating a test dashboard

Kobzol committed:
Add buttons for expanding and collapsing all test suites

Kobzol committed:
Add a note about how to find tests that haven't been executed anywhere.

Kobzol committed:
Render test revisions separately

Kobzol committed:
Create a macro for rendering test results

Kobzol committed:
Print number of root tests and subdirectories

Kobzol committed:
Turn `test_dashboard` into a file

Kobzol committed:
Add a note about the test dashboard to the post-merge report

jieyouxu commented:
FTR, the dashboard currently looks like:

![Screenshot 2025-04-18 164349](https://github.com/user-attachments/assets/8a54a084-4032-42cd-b746-526bbbc65762)
![Screenshot 2025-04-18 164433](https://github.com/user-attachments/assets/45f141d6-fb0c-4d34-b40c-55ef348fb5e8)

jieyouxu commented on src/ci/citool/Cargo.toml:
```diff
@@ -5,6 +5,7 @@ edition = "2021"
 
 [dependencies]
 anyhow = "1"
+askama = "0.13"
```
Discussion: should this be gated behind a feature, or is the intention that the dashboard might be usable even in CI? Or rather, is it even worth the maintenance effort to gate it behind a feature?

Kobzol replied:
I think that eventually we might want to run it also on CI and then upload the generated website to S3 or something, to just provide a link in the post-merge report, so that users don't have to run it locally.

I wanted to feature gate advanced dependencies (https://github.com/rust-lang/rust/pull/139678), but we decided to use a cache on CI instead to make the build faster.

jieyouxu replied:
Fair enough, that sounds like a good plan

jieyouxu commented on src/ci/citool/src/main.rs:
````diff
@@ -188,6 +188,20 @@ fn post_merge_report(db: JobDatabase, current: String, parent: String) -> anyhow
     let mut job_info_resolver = JobInfoResolver::new();
     output_test_diffs(&metrics, &mut job_info_resolver);
+
+    output_details("Test dashboard", || {
+        println!(
+            r#"\nRun
+
+```bash
+cargo run --manifest-path src/ci/citool/Cargo.toml -- \
+    test-dashboard {current} --output-dir test-dashboard
+```
+And then open `test-dashboard/index.html` in your browser to see an overview of all executed tests.
+"#
+        );
+    });
````
Suggestion: could you also include this bit in the citool README.md?

Kobzol replied:
This is shown in the post-merge report, which is a comment sent on every merged PR.

jieyouxu replied:
Ah right, I forgor about that

jieyouxu reviewed (approved):
Thanks for working on this! The impl looks good, played around with the generated dashboard locally and the initial version's content looks like what I'd expect.

jieyouxu commented:
An example test (edited test path):

```
tests/codegen/autodiff/scalar.rs (passed: 0, ignored: 23)
```

That seems a bit surprising -- `passed: 0` looks right, I definitely don't satisfy `//@ needs-enzyme` locally. However the `ignored: 23` seems a bit high -- this is a codegen test that doesn't use any revisions 🤔

Kobzol commented:
The number is a cartesian product of `stage x target x CI job` combinations where the test was executed or ignored. In most cases, there is 1 stage and 1 target per CI job, so it's more or less the number of CI jobs that ran the test.

jieyouxu commented:
Hm right.

jieyouxu commented:
> The number is a cartesian product of `stage x target x CI job` combinations where the test was executed or ignored. In most cases, there is 1 stage and 1 target per CI job, so it's more or less the number of CI jobs that ran the test.

Should we maybe add this as like a banner thing on the page, because I imagine I won't be the first one to ask this question :D

Kobzol commented:
Yeah, I would like to extend the amount of information on the page, but the problem is that with 20k tests every small amount of HTML per test is gonna add up quickly :laughing: The page already has 2 MiB. I wanted to land this as an initial version and then do incremental improvements in follow-up PRs, to avoid having too much stuff at once in a PR.

jieyouxu commented:
Oh I just mean an *overall* note, like

> Here's how to interpret the "passed" and "ignored" counts: they are coalsced between stage x target x CI jobs where the test was executed or ignored. A test is never run across all CI jobs if its "passed" count is 0.

Not like, individual per-test remarks, if that makes sense?

jieyouxu commented:
But in any case I don't think it's worth blocking over that, so feel free to r=me with or without the remark.

jieyouxu commented:
Thanks!
@bors r+ rollup

bors commented:
:pushpin: Commit cecf16785f193c247c47e42524322aeccc96c8a4 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu cecf16785f193c247c47e42524322aeccc96c8a4 -->
<!-- homu: {"type":"Approved","sha":"cecf16785f193c247c47e42524322aeccc96c8a4","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
I was just about to add the note =D I also noticed one thing I want to fix.

@bors r-

Kobzol committed:
Add a note that explains the counts

Kobzol committed:
Reduce duplicated test prefixes in nested subdirectories

`assembly/asm` contained a test named `asm/aarch64-el2vmsa.rs`, while it should have been only `arch64-el2vmsa.rs`.

Kobzol commented:
Added the note and removed duplicated subdirectory name in the name of nested tests.

jieyouxu reviewed (approved):
Looks good, thanks. r=me after PR CI is 🍏

Kobzol commented:
@bors r=jieyouxu rollup

bors commented:
:pushpin: Commit b18e37305304780530224736ad55145063c7e8a6 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu b18e37305304780530224736ad55145063c7e8a6 -->
<!-- homu: {"type":"Approved","sha":"b18e37305304780530224736ad55145063c7e8a6","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
